File: src/main/java/com/portofino/realtrainmodunofficial/rail/util/Point.java
package com.portofino.realtrainmodunofficial.rail.util;

import net.minecraft.world.level.Level;

public final class Point {
    private static final int MAX_COUNT = 80;

    public final RailPosition rpRoot;
    public final RailMapSwitch rmMain;
    public final RailMapSwitch rmBranch;
    public final RailDir branchDir;
    public final boolean mainDirIsPositive;
    public final boolean branchDirIsPositive;
    private int moveCount;

    public Point(RailPosition railPos, RailMapSwitch rms1, RailMapSwitch rms2) {
        this.rpRoot = railPos;
        boolean mainFirst = rms1.getLength() <= rms2.getLength();
        this.rmMain = mainFirst ? rms1 : rms2;
        this.rmBranch = mainFirst ? rms2 : rms1;
        this.branchDir = getDir(this.rpRoot, this.rmMain, this.rmBranch);
        this.mainDirIsPositive = this.rmMain.getStartRP() == this.rpRoot;
        this.branchDirIsPositive = this.rmBranch.getStartRP() == this.rpRoot;
    }

    public Point(RailPosition railPos, RailMapSwitch rms1) {
        this.rpRoot = railPos;
        this.rmMain = rms1;
        this.rmBranch = null;
        this.branchDir = RailDir.NONE;
        this.mainDirIsPositive = rms1.getStartRP() == railPos;
        this.branchDirIsPositive = false;
    }

    private static RailDir getDir(RailPosition root, RailMapSwitch rms1, RailMapSwitch rms2) {
        RailPosition rp1 = rms1.getStartRP() == root ? rms1.getEndRP() : rms1.getStartRP();
        RailPosition rp2 = rms2.getStartRP() == root ? rms2.getEndRP() : rms2.getStartRP();
        return root.getDir(rp1, rp2);
    }

    public void onUpdate(Level level) {
        if (level == null) {
            return;
        }
        boolean powered = this.rpRoot.checkRSInput(level);
        if (powered) {
            if (this.moveCount < MAX_COUNT) {
                ++this.moveCount;
            }
        } else if (this.moveCount > 0) {
            --this.moveCount;
        }
    }

    public float getMovement() {
        return (float) this.moveCount / (float) MAX_COUNT;
    }

    public RailMap getActiveRailMap(Level level) {
        if (this.branchDir == RailDir.NONE || this.rmBranch == null) {
            return this.rmMain;
        }
        return this.rpRoot.checkRSInput(level) ? this.rmBranch : this.rmMain;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/rail/util/Point.kt
package cc.mirukuneko.realtrainmodrenewed.rail.util

import net.minecraft.world.level.Level

class Point {
    @JvmField
    val rpRoot: RailPosition

    @JvmField
    val rmMain: RailMapSwitch

    @JvmField
    val rmBranch: RailMapSwitch?

    @JvmField
    val branchDir: RailDir

    @JvmField
    val mainDirIsPositive: Boolean

    @JvmField
    val branchDirIsPositive: Boolean

    private var moveCount = 0

    constructor(railPos: RailPosition, rms1: RailMapSwitch, rms2: RailMapSwitch) {
        rpRoot = railPos
        val mainFirst = rms1.length <= rms2.length
        rmMain = if (mainFirst) rms1 else rms2
        rmBranch = if (mainFirst) rms2 else rms1
        branchDir = getDir(rpRoot, rmMain, rmBranch)
        mainDirIsPositive = rmMain.startRP === rpRoot
        branchDirIsPositive = rmBranch.startRP === rpRoot
    }

    constructor(railPos: RailPosition, rms1: RailMapSwitch) {
        rpRoot = railPos
        rmMain = rms1
        rmBranch = null
        branchDir = RailDir.NONE
        mainDirIsPositive = rms1.startRP === railPos
        branchDirIsPositive = false
    }

    fun onUpdate(level: Level?) {
        if (level == null) {
            return
        }
        if (rpRoot.checkRSInput(level)) {
            if (moveCount < MAX_COUNT) {
                ++moveCount
            }
        } else if (moveCount > 0) {
            --moveCount
        }
    }

    fun getMovement(): Float = moveCount.toFloat() / MAX_COUNT.toFloat()

    fun getActiveRailMap(level: Level): RailMap {
        return if (branchDir == RailDir.NONE || rmBranch == null) {
            rmMain
        } else if (rpRoot.checkRSInput(level)) {
            rmBranch
        } else {
            rmMain
        }
    }

    companion object {
        private const val MAX_COUNT = 80

        private fun getDir(root: RailPosition, rms1: RailMapSwitch, rms2: RailMapSwitch): RailDir {
            val rp1 = if (rms1.startRP === root) rms1.endRP else rms1.startRP
            val rp2 = if (rms2.startRP === root) rms2.endRP else rms2.startRP
            return root.getDir(rp1, rp2)
        }
    }
}
